File: new-JS-applets/JS Files/Kollisionen/HashTabelle.js
(function(){var P$=Clazz.newPackage("Kollisionen"),I$=[];
/*c*/var C$=Clazz.newClass(P$, "HashTabelle", null, 'java.util.Hashtable');

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
this.maxhoehe=19;
},1);

C$.$fields$=[['I',['maxhoehe'],'O',['saeule','int[]']]
,['I',['hsize','dszahl'],'O',['numbers','java.util.Hashtable','wert','Integer']]]

Clazz.newMeth(C$, 'c$', function () {
Clazz.super_(C$, this);
C$.hsize=1;
this.saeule=Clazz.array(Integer.TYPE, [C$.hsize]);
}, 1);

Clazz.newMeth(C$, 'c$$I', function (felder) {
Clazz.super_(C$, this);
C$.hsize=felder;
this.saeule=Clazz.array(Integer.TYPE, [C$.hsize]);
}, 1);

Clazz.newMeth(C$, 'HashEintrag$I$I', function (datensatz, methindex) {
var hashwert;
var ds= new Integer(datensatz);
if (methindex == 1) hashwert=C$.HashCode1$O(ds).intValue$();
 else if (methindex == 2) hashwert=C$.HashCode2$O(ds).intValue$();
 else if (methindex == 3) hashwert=C$.HashCode3$O(ds).intValue$();
 else if (methindex == 4) hashwert=C$.HashCode4$O(ds).intValue$();
 else if (methindex == 5) hashwert=C$.HashCode5$O(ds).intValue$();
 else hashwert=C$.HashCode5$O(ds).intValue$();
System.out.println$S("Hashwert: " + hashwert);
return hashwert;
});

Clazz.newMeth(C$, 'NeueSaeule$I', function (feldeintrag) {
if (this.saeule[feldeintrag] < 19) this.saeule[feldeintrag]+=1;
});

Clazz.newMeth(C$, 'HashCode1$O', function (obj) {
var summe;
C$.dszahl=(obj).intValue$();
summe=(C$.dszahl % 10);
C$.wert= new Integer(summe % C$.hsize);
return C$.wert;
}, 1);

Clazz.newMeth(C$, 'HashCode2$O', function (obj) {
var summe;
var zwwert;
C$.dszahl=(obj).intValue$();
summe=0;
zwwert=0;
if (C$.dszahl > 999) {
zwwert=((C$.dszahl - (C$.dszahl % 1000))/1000|0);
} else zwwert=0;
summe+=zwwert;
summe+=(C$.dszahl % 10);
C$.wert= new Integer(summe % C$.hsize);
return C$.wert;
}, 1);

Clazz.newMeth(C$, 'HashCode3$O', function (obj) {
var summe;
C$.dszahl=(obj).intValue$();
summe=0;
do {
var zwwert=0;
if (C$.dszahl > 999) {
zwwert=((C$.dszahl - (C$.dszahl % 1000))/1000|0);
C$.dszahl=C$.dszahl % 1000;
} else if (C$.dszahl > 99) {
zwwert=((C$.dszahl - (C$.dszahl % 100))/100|0);
C$.dszahl=C$.dszahl % 100;
} else if (C$.dszahl > 9) {
zwwert=((C$.dszahl - (C$.dszahl % 10))/10|0);
C$.dszahl=C$.dszahl % 10;
} else if (C$.dszahl > 0) {
zwwert=C$.dszahl;
C$.dszahl=0;
}summe=summe + zwwert;
} while (C$.dszahl != 0);
C$.wert= new Integer(summe % C$.hsize);
return C$.wert;
}, 1);

Clazz.newMeth(C$, 'HashCode4$O', function (obj) {
var rest;
C$.dszahl=(obj).intValue$();
if (C$.hsize <= C$.dszahl) {
rest=(C$.dszahl % C$.hsize);
} else rest=C$.dszahl;
C$.wert= new Integer(rest);
return C$.wert;
}, 1);

Clazz.newMeth(C$, 'HashCode5$O', function (obj) {
var rest;
var teta;
var kleinezahl;
var zwwert;
C$.dszahl=(obj).intValue$();
teta=(Math.sqrt(5.0) - 1) / 2;
zwwert=(((C$.dszahl * teta)|0));
kleinezahl=Math.abs(zwwert - (C$.dszahl * teta));
rest=Math.ceil((C$.hsize * kleinezahl) - 1);
C$.wert= new Integer((rest|0));
return C$.wert;
}, 1);
})();
;Clazz.setTVer('3.2.9-v1');//Created 2021-01-28 11:37:28 Java2ScriptVisitor version 3.2.9-v1 net.sf.j2s.core.jar version 3.2.9-v1

File: src/old/Kollisionen/HashTabelle.java
package old.Kollisionen;
import java.util.*;
import java.lang.*;

public class HashTabelle extends Hashtable {

    private static Hashtable numbers;
    private static int hsize;
    private static Integer wert;
    private static int dszahl;
	private final int maxhoehe = 19;
    int[] saeule;

    // Konstruktoren 

    HashTabelle() {
	// Bei Bedarf wird auf Klasse java.util.Hashtable zugegriffen. 
	// numbers = new Hashtable(1);
	hsize = 1;
	saeule = new int[hsize];
    }    

    HashTabelle(int felder) 
	{
	    // Bei Bedarf wird auf Klasse java.util.Hashtable zugegriffen.
	    // numbers = new Hashtable(felder);
	    hsize = felder;
	    saeule = new int[hsize];
	}
    

    public int HashEintrag(int datensatz, int methindex) {//ermittelt Hashwert

	int hashwert;
	Integer ds = new Integer(datensatz);

	// Zuweisung des Hashwertes durch die Methode HashCodex(); 
	if (methindex==1)
	    hashwert = HashCode1(ds).intValue();
	else if (methindex==2)
	    hashwert = HashCode2(ds).intValue();
	else if (methindex==3)
	    hashwert = HashCode3(ds).intValue();
	else if (methindex==4)
	    hashwert = HashCode4(ds).intValue();
	else if (methindex==5)
	    hashwert = HashCode5(ds).intValue();
	else hashwert = HashCode5(ds).intValue();
	System.out.println("Hashwert: "+hashwert);

	// falls gewünscht, muß number im Konstruktor inst. worden sein.
	// this.numbers.put(ds, new Integer(hashwert));

	return hashwert;	
    } // HashEintrag(int datensatz, int methindex) 


    public void NeueSaeule(int feldeintrag) { // Säulenindex wird aufdatiert
	if (this.saeule[feldeintrag] < maxhoehe)
	    this.saeule[feldeintrag] += 1;
    }

    private static Integer HashCode1(Object obj) { // 1. Divisionsmethode
	
	// Diese Methode ist die 1. Divisionsmethode;
	// Hashwert wird mit letztem Zeichen berechnet (eher schlecht).
	
	int summe;
	
	dszahl = ((Integer)obj).intValue();
	
	summe = (dszahl%10);
	    
	    wert = new Integer(summe % hsize);
	return wert;
    } // Ende HashCode1(Object obj); 
    


    private static Integer HashCode2(Object obj) { // 2. Divisionsmethode
	
	// Diese Methode ist die 2. Divisionsmethode;
	// Sie summiert das erste und letzte Zeichen des Int.-Wertes, bevor   
	// per Division der Hashwert gebildet wird (etwas besser).

	int summe;
	int zwwert;

	dszahl = ((Integer)obj).intValue();
	
	summe = 0;
	zwwert = 0;
	    if (dszahl > 999) { // 1000er Ziffer wird evtl. ermittelt
		zwwert = (dszahl-(dszahl%1000))/1000;
	    }
	    else // Das erste Zeichen ist sonst eine "0"
		zwwert = 0;
	    summe += zwwert;
	    // 1er Ziffer wird ermittelt 
	    summe += (dszahl%10);
	    
	    wert = new Integer(summe % hsize);
	return wert;
    } // Ende HashCode2(Object obj); 


    private static Integer HashCode3(Object obj) { // 3. Divisionsmethode

	// Diese Methode ist die 3. Divisionsmethode;
	// Die Quersumme des Integer-Datensatzes wird gebildet und 
	// anschliessend per Division der Hashwert ermittelt (noch besser).

	int summe;
	
	dszahl = ((Integer)obj).intValue();
	
	// In der folgenden do-while-Schleife wird der Int.-wert pro Iteration 
	// um eine 10er Potenz reduziert, bevor die entspr. Ziffer aufsummiert 
	// wird. Ultrastabil!
	
	summe = 0;
	do {
	    int zwwert = 0;
	    if (dszahl > 999) { // 1000er Ziffer wird ermittelt
		zwwert = (dszahl-(dszahl%1000))/1000;
		dszahl = dszahl%1000;
	    }
	    else if (dszahl > 99) { // 100er Ziffer wird ermittelt
		zwwert = (dszahl-(dszahl%100))/100;
		dszahl = dszahl%100;
	    }
	    else if (dszahl > 9) { // 10er Ziffer wird ermittelt
		zwwert = (dszahl-(dszahl%10))/10;
		dszahl = dszahl%10;
	    }
	    else if (dszahl > 0) { // 1er Ziffer wird ermittelt
		zwwert = dszahl;
		dszahl = 0;
	    }
	    summe = summe + zwwert;
	
	} while (dszahl!=0);
	
	wert = new Integer(summe % hsize);
	return wert;
    } // Ende HashCode3(Object obj); 


    private static Integer HashCode4(Object obj) { // 4. Divisionsmethode

	// Diese Methode berechnet (eingabezahl mod(hsize)); dies ist 
	// natürlich nur mit Integer-Eingaben möglich. Daher nicht klassisch.
	// I.A. noch besser als Methode 3 aufgrund der Streuung.

	int rest;
	dszahl = ((Integer)obj).intValue();	
	if (hsize <= dszahl) {
	    rest = (dszahl % hsize);
	}
	else rest = dszahl;
	wert = new Integer(rest);
	return wert;
    } // Ende HashCode4(Object obj); 


    private static Integer HashCode5(Object obj) { // 5. Methode (Multipl.)
	// Diese Methode ist eine Multiplikationsmethode (beste Streuung) 
	// Integer-Datensatz wird mit geeignetet Faktor (--> teta) multipl.,
	// anschl. die Nachkommastelle betrachtet => aus (0,1) anschl. wie 
	// bei "random" auf die Anzahl der Felder verteilt!
	     
	double rest;
	double teta;
	double kleinezahl;
	int zwwert;
	
	dszahl = ((Integer)obj).intValue();
	teta = (Math.sqrt(5.0)-1)/2;
	
	zwwert = ((int)(dszahl * teta));
	kleinezahl = Math.abs(zwwert-(dszahl*teta));
	rest = Math.ceil((hsize * kleinezahl)-1);
	
	wert = new Integer((int)rest);
	return wert;
    } // Ende HashCode5(Object obj); 
    
    
} // Ende class HashTabelle extends Hashtable







